### PR TITLE
New dashboard

### PR DESCRIPTION
**Pull request by paweltin**
**Product list**
- [x] give grey color to price.
- [x] remove delete link
- [x] fixed FAB
- [x] add button redirects to product form directly

**Product form**
- [x] current 'add' buttons on product Variants and Stock tabs look like submits - we should try putting them in Data Table's header.
- [x] product attributes like Weight should have units displayed as suffixes (e.g. Price).
- [x] add title attribute on product image description
- [x] add delete link (in header dropdown)
- [x] improve file input widget
- [x] delete confirmation as modal window
- [x] make sure that non-field errors are displayed (e.g. in stock form, when trying to add a stock item that already exists)

**Categories**
- [x] categories should list only one level
- [x] adding a child is possible only on parent's page
- [x] parent is editable

**Common**
- [x] add breadcrumbs everywhere.
- [x] breadcrumbs in header
- [x] add native back button on mobile
- [x] use new Material Design Icons
- [x] logo as SVG
- [x] group menu items
- [x] unify form submit messages (snackbars?)

### CODE REVIEW

File: saleor/dashboard/category/forms.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class CategoryForm(forms.ModelForm):
+    class Meta:
+        model = Category
+        exclude = []
+
+    def clean_parent(self):
+        parent = self.cleaned_data['parent']
+        if parent == self.instance:
```
Comment (patrys) [inline saleor/dashboard/category/forms.py]: That won't suffice. You need to check whether the current object is not within its new parents ancestors. Loops in tree structures are not allowed.

**Comment by patrys** [comment]
The sliding labels are confusing at best, especially on desktop. It's hard to distinguish fields from each other and the font changing size depending on position makes empty fields look like sub-headers. Can we get rid of them and add example placeholder text where it makes sense?

**Comment by paweltin** [comment]
Current 'add' buttons on product Variants and Stock tabs look like submits. We should try putting them in Data Table's header.
Product attributes like Weight should have units displayed as suffixes (e.g. Price).
We should put table filters in Data Table's header.
Categories should list only one level; adding a child is possible only on parent's page (parent is editable); add drag'n'drop feature.
Add collection placeholder on product list, give grey color to price.
Add Breadcrumbs everywhere.

File: saleor/dashboard/category/forms.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class CategoryForm(forms.ModelForm):
+    class Meta:
+        model = Category
+        exclude = []
+
+    def __init__(self, *args, **kwargs):
+        super(CategoryForm, self).__init__(*args, **kwargs)
+        self.fields['parent'] = TreeNodeChoiceField(queryset=Category.objects.all(), required=False)
```
Comment (patrys) [inline saleor/dashboard/category/forms.py]: If the field does not depend on the current instance, why is it created in `__init__`?

File: saleor/dashboard/category/forms.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+from mptt.forms import TreeNodeChoiceField
+
+from ...product.models import Category
+
+
+class CategoryForm(forms.ModelForm):
```
Comment (patrys) [inline saleor/dashboard/category/forms.py]: From the code it seems that this form is used for editing existing categories as well as creating new ones. If so, you may want to use `category.move_to(…)` or extend `mptt.forms.MoveNodeForm`.

File: saleor/dashboard/category/urls.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+from django.conf.urls import patterns, url
+
+from . import views
+
+
+urlpatterns = patterns(
```
Comment (patrys) [inline saleor/dashboard/category/urls.py]: `urlpatterns` can be a regular list now. There is no need to call `patterns(…)`.

File: saleor/dashboard/category/views.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+@staff_member_required
+def category_list(request, root=None):
+    ctx = {}
+    categories = Category.objects.all()
+    if root:
+        current_node = get_object_or_404(Category, pk=root)
+        categories = current_node.get_descendants()
+        ctx['current_node'] = current_node
+        ctx['category_breadcrumbs'] = current_node.get_ancestors(include_self=True)
+    min_level = categories[0].get_level() if categories else 0
```
Comment (patrys) [inline saleor/dashboard/category/views.py]: What is this sorcery? `… if categories …` evaluates the entire queryset. That is not the best idea.

Definitely not the best

File: saleor/dashboard/category/views.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    ctx['min_level'] = min_level
+    return TemplateResponse(request, 'dashboard/category/list.html', ctx)
+
+
+@staff_member_required
+def category_details(request, pk=None, parent_pk=None):
+    ctx = {}
+    if pk:
+        category = get_object_or_404(Category.objects.all(), pk=pk)
+        title = category.name
```
Comment (patrys) [inline saleor/dashboard/category/views.py]: Isn't this the template's responsibility to handle? Changing UI strings should not require me to alter backend code.

File: saleor/dashboard/category/views.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    else:
+        category = Category()
+        title = _('Add new category')
+    initial = {}
+    if parent_pk:
+        initial['parent'] = parent_pk
+        ctx['parent_pk'] = parent_pk
```
Comment (patrys) [inline saleor/dashboard/category/views.py]: Why is a PK needed in the template?

File: saleor/dashboard/category/views.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    min_level = categories[0].get_level() if categories else 0
+    max_level = min_level + 1
+    categories = categories.filter(level__gte=min_level, level__lte=max_level)
+    ctx['categories'] = categories
+    ctx['min_level'] = min_level
+    return TemplateResponse(request, 'dashboard/category/list.html', ctx)
+
+
+@staff_member_required
+def category_details(request, pk=None, parent_pk=None):
```
Comment (patrys) [inline saleor/dashboard/category/views.py]: How about `category_details(request, category_id, parent_category_id)`? Why is parent category needed here? Can't we have a separate `create_category` view for that purpose?

File: saleor/dashboard/category/views.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        if form.errors:
+            messages.error(request, _('Failed to save category'))
+    ctx.update({'category': category, 'form': form, 'title': title})
+    return TemplateResponse(request, 'dashboard/category/detail.html', ctx)
+
+
+class CategoryDeleteView(StaffMemberOnlyMixin, DeleteView):
+    model = Category
+    template_name = 'dashboard/category/category_confirm_delete.html'
+    success_url = reverse_lazy('dashboard:categories')
```
Comment (patrys) [inline saleor/dashboard/category/views.py]: Deleting a sub-sub-category should not result in you getting redirected to the root list.

File: saleor/dashboard/product/forms.py [hunk 10]
```diff
@@ -2,77 +2,161 @@
...
 
 
-class ShirtForm(ProductForm):
+class StockForm(forms.ModelForm):
     class Meta:
-        model = Shirt
+        model = Stock
         exclude = []
+        widgets = {
+            'product': forms.HiddenInput()
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: Why is there a `product` field in the form? What if I intentionally alter the hidden value?

File: saleor/dashboard/product/forms.py [hunk 11]
```diff
@@ -2,77 +2,161 @@
...
+        }
 
+    def __init__(self, *args, **kwargs):
+        product = kwargs.pop('product')
+        super(StockForm, self).__init__(*args, **kwargs)
+        variants = product.variants.all()
+        if variants:
+            self.fields['variant'].choices = [(variant.pk, variant) for variant
+                                              in variants]
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: Isn't that exactly what a `ModelChoiceField` does?

File: saleor/dashboard/product/forms.py [hunk 12]
```diff
@@ -2,77 +2,161 @@
...
+
 
-class ImageInputWidget(ClearableFileInput):
-    url_markup_template = '<a href="{0}"><img src="{0}" width=50 /></a>'
+class ProductVariantForm(forms.ModelForm):
+    class Meta:
+        model = ProductVariant
+        exclude = ['attributes']
+        widgets = {
+            'product': forms.HiddenInput(),
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: Why is there a `product` field here? Can I move a variant to a different product?

Comment (patrys) [inline saleor/dashboard/product/forms.py]: Also trailing comma and a dangling curly brace.

File: saleor/dashboard/product/forms.py [hunk 13]
```diff
@@ -2,77 +2,161 @@
...
+    if isinstance(product, Product):
+        return ProductForm
     else:
-        raise ValueError('Unknown product')
+        raise ValueError('Unknown product class')
 
 
 def get_product_cls_by_name(cls_name):
-    if not cls_name in PRODUCT_CLASSES:
+    for cls in PRODUCT_CLASSES.keys():
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: `foo in dict_a` and `foo in dict_a.keys()` are equivalent.

File: saleor/dashboard/product/forms.py [hunk 14]
```diff
@@ -2,77 +2,161 @@
...
+    class Meta:
+        model = ProductAttribute
+        exclude = ['slug']
+
+    def __init__(self, *args, **kwargs):
+        super(ProductAttributeForm, self).__init__(*args, **kwargs)
+
+    def clean(self):
+        cleaned_data = super(ProductAttributeForm, self).clean()
+        slug = slugify(self.cleaned_data['display'])
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: This makes it possible to accidentally change the slug when correcting a typo. Always use [Unidecode](https://pypi.python.org/pypi/Unidecode) before feeding unicode data to `slugify(…)` as it breaks lots of diacritics.

File: saleor/dashboard/product/forms.py [hunk 15]
```diff
@@ -2,77 +2,161 @@
...
+        if ProductAttribute.objects.filter(slug=slug):
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: `.exists()`.

File: saleor/dashboard/product/forms.py [hunk 16]
```diff
@@ -2,77 +2,161 @@
...
+            raise forms.ValidationError(
+                pgettext_lazy('Validation error', 'Attribute already exists'))
```
Comment (patrys) [inline saleor/dashboard/product/forms.py]: You use an automatically generated field to enforce uniqueness but the user has no way to correct said field if two different display names result in the same slug being generated.

File: saleor/dashboard/product/views.py [hunk 17]
```diff
@@ -2,66 +2,83 @@
...
 from django.utils.translation import ugettext_lazy as _
 from django.views.generic import DeleteView
 
-from ...product.models import Product
+from ...product.models import Product, ProductImage, Stock, ProductAttribute
 from ..utils import paginate
 from ..views import StaffMemberOnlyMixin, staff_member_required
-from .forms import (ProductImageFormSet, ProductClassForm, get_product_form,
-                    get_product_cls_by_name, get_variant_formset)
+from .forms import (ProductClassForm, get_product_form,
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Can we switch to `from . import forms`?

File: saleor/dashboard/product/views.py [hunk 18]
```diff
@@ -2,66 +2,83 @@
...
+                    get_product_cls_by_name, get_variant_form,
+                    ProductImageForm, get_verbose_name, StockForm,
+                    get_variant_cls, ProductAttributeForm,
+                    AttributeChoiceValueFormset, VariantAttributesForm)
 
 
 @staff_member_required
 def product_list(request):
-    products = Product.objects.prefetch_related('images').select_subclasses()
+    products = Product.objects.prefetch_related('images')
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Why do we drop `select_subclasses`? Can I no longer have more than one product class?

File: saleor/dashboard/product/views.py [hunk 19]
```diff
@@ -2,66 +2,83 @@
...
 
 @staff_member_required
-def product_details(request, pk=None, cls_name=None):
-    if pk:
-        product = get_object_or_404(Product.objects.select_subclasses(), pk=pk)
-        title = product.name
+def product_details(request, pk=None, product_cls=None):
+    creating = pk is None
+    initial = {}
+    if creating:
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: I consider this to be an anti-pattern. Create a separate `create_product` view and keep the logic clean and separate.

File: saleor/dashboard/product/views.py [hunk 20]
```diff
@@ -2,66 +2,83 @@
...
+        if creating and product:
+            return redirect('dashboard:product-update', pk=product.pk)
+
+    ctx = {'title': title, 'product': product, 'attributes': attributes,
+           'images': images, 'product_form': form, 'stock_items': stock_items,
+           'variants': variants}
+    if pk:
+        images_reorder_url = reverse_lazy('dashboard:product-images-reorder',
+                                          kwargs={'product_pk': pk})
+        ctx['images_reorder_url'] = images_reorder_url
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Why is this URL constructed here and not in the template?

File: saleor/dashboard/product/views.py [hunk 21]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
         messages.success(request, _('Deleted product %s') % self.object)
         return result
+
+
+@staff_member_required
+def stock_edit(request, product_pk, stock_pk=None):
+    product = get_object_or_404(Product, pk=product_pk)
+    if stock_pk:
+        stock = get_object_or_404(Stock, pk=stock_pk)
+        title = stock.variant.name
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: This belongs in templates as it's display logic.

File: saleor/dashboard/product/views.py [hunk 22]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
+        title = _('Add variant')
+
+    form_cls = get_variant_form(product)
+    form = form_cls(request.POST or None, instance=variant,
+                    initial=form_initial)
+    attributes_form = VariantAttributesForm(request.POST or None,
+                                            instance=variant)
+    forms = [form, attributes_form]
+
+    if all([f.is_valid() for f in forms]):
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: `if all([form.is_valid(), attributes_form.is_valid()]):`

File: saleor/dashboard/product/views.py [hunk 23]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
+                                    pk=variant_pk)
+        title = variant.name
+    else:
+        variant = variant_cls(product=product)
+        title = _('Add variant')
+
+    form_cls = get_variant_form(product)
+    form = form_cls(request.POST or None, instance=variant,
+                    initial=form_initial)
+    attributes_form = VariantAttributesForm(request.POST or None,
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: `attribute_form`, `VariantAttributeForm` should be singular just like you say "file manager" and not "files manager".

File: saleor/dashboard/product/views.py [hunk 24]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
+
+    if all([f.is_valid() for f in forms]):
+        form.save()
+        attributes_form.save()
+        if variant_pk:
+            msg = _('Updated variant %s') % variant.name
+        else:
+            msg = _('Added variant %s') % variant.name
+        messages.success(request, msg)
+        success_url = request.POST['success_url']
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Sanitization required.

File: saleor/dashboard/product/views.py [hunk 25]
```diff
@@ -2,66 +2,83 @@
...
             messages.error(request, _('Your submitted data was not valid - '
-                           'please correct the errors below'))
-    ctx = {'title': title, 'product': product}
-    ctx.update(forms)
-    return TemplateResponse(request, 'dashboard/product/product_form.html', ctx)
+                                      'please correct the errors below'))
+    return product
 
 
 class ProductDeleteView(StaffMemberOnlyMixin, DeleteView):
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Can we make this a regular function to match the views for deleting variants and stock?

File: saleor/dashboard/product/views.py [hunk 26]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
+    return TemplateResponse(request, 'dashboard/product/attributes/list.html',
+                            ctx)
+
+
+@staff_member_required
+def attribute_edit(request, pk=None):
+    if pk:
+        attribute = get_object_or_404(ProductAttribute, pk=pk)
+        title = attribute.display
+    else:
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: Again, creation and edition need separate views.

File: saleor/dashboard/product/views.py [hunk 27]
```diff
@@ -73,3 +90,193 @@ def post(self, request, *args, **kwargs):
...
+        attribute = ProductAttribute()
+        title = _('Add new attribute')
+    form = ProductAttributeForm(request.POST or None, instance=attribute)
+    formset = AttributeChoiceValueFormset(request.POST or None,
+                                          request.FILES or None,
+                                          instance=attribute)
+    if form.is_valid() and formset.is_valid():
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: When the main form is not valid, the formset is not validated at all.

File: saleor/dashboard/views.py [hunk 28]
```diff
@@ -65,6 +61,5 @@ def index(request):
 def get_low_stock_products():
     threshold = getattr(settings, 'LOW_STOCK_THRESHOLD', 10)
     products = Product.objects.filter(
-        Q(shirt__variants__stock__lte=threshold) |
-        Q(bag__variants__stock__lte=threshold)).distinct()
+        Q(variants__stock__quantity__lte=threshold)).distinct()
```
Comment (patrys) [inline saleor/dashboard/views.py]: This is more tricky as you should check only the highest stock level, not all of them. If one location contains 100 of the item, then the other only having 1 is not a reason to panic.

File: bower.json [hunk 29]
```diff
@@ -10,6 +10,11 @@
     "less": "~2.4.0",
     "bootstrap": "~3.3.2",
     "components-font-awesome": "~4.3.0",
-    "zocial-less": "~1.0.0"
+    "zocial-less": "~1.0.0",
+    "materialize": "~0.97.0",
+    "jquery": "~2.1.4",
+    "Sortable": "git@github.com:RubaXa/Sortable.git#~1.2.0",
```
Comment (patrys) [inline bower.json]: Isn't it available outside of GitHub?

**Comment by mirekm** [comment]
:zap: :tada:

**Comment by hongquan** [comment]
The new dashboard is missing file _/static/dashboard-head.js_ (error 404)

**Pull request merged**